net/instaweb/http/redirect_following_url_async_fetcher.cc

```diff
@@ -0,0 +1,333 @@
+      return false;
+    }
+    return true;
+  }
+
+  bool TryMapRedirect(const GoogleString& redirect_url,
+                           GoogleString& mapped_url) {
+
+    const RewriteOptions* options =
+        redirect_following_fetcher_->rewrite_options();
+    const DomainLawyer* domain_lawyer = options->domain_lawyer();
+
+    bool cacheable = response_headers()->IsProxyCacheable(
+        base_fetch_->request_headers()->GetProperties(),
+        ResponseHeaders::GetVaryOption(options->respect_vary()),
+        ResponseHeaders::kNoValidator);
+
+    if (!cacheable) {
+      EmitRedirectWarning(url_, "t.b.d.", "Redirect not cacheable, not following");
+      return false;
+    } else {
+      max_age_ = std::min(max_age_, response_headers()->cache_ttl_ms());
+    }
+    GoogleUrl redirect_gurl(gurl_, redirect_url);
+    if (!domain_lawyer->IsDomainAuthorized(GoogleUrl(context_url_),
+                                           redirect_gurl)) {
+      EmitRedirectWarning(context_url_, redirect_url, "Unauthorized");
+      return false;
+    }
+
+    if (!options->IsAllowed(redirect_gurl.Spec())) {
+      EmitRedirectWarning(context_url_, redirect_url, "Rewriting disallowed");
+      return false;
+    }
+
+    GoogleString mapped_domain_name;
+    GoogleString host_header;
+    bool is_proxy;
+    bool mapped = domain_lawyer->MapOriginUrl(
+        redirect_gurl, &mapped_domain_name, &host_header, &is_proxy);
+
+    if (mapped) {
+      redirect_gurl.Reset(mapped_domain_name);
+      mapped_url.assign(mapped_domain_name);
+      if (redirect_gurl.SchemeIs("https") &&
+          !redirect_following_fetcher_->SupportsHttps()) {
+        EmitRedirectWarning(url_, redirect_url, "Https not supported");
+        return false;
+      }
+      if (!is_proxy) {
+        request_headers()->Replace(HttpAttributes::kHost, host_header);
+      }
+    } else {
+      // Shouldn't happen
+      DCHECK(false);
+      EmitRedirectWarning(url_, redirect_url, "Invalid mapped url");
+      return false;
+    }
+    return true;
+  }
+
+  bool TryExtractRedirectUrlFromResponseHeaders(GoogleString& redirect_url) {
+    const char* raw_redirect_url =
+        response_headers()->Lookup1(HttpAttributes::kLocation);
+
+    if (!raw_redirect_url) {
+      EmitRedirectWarning(
+          url_, "none", "Failed looking up exactly one Location header");
+      return false;
+    } else if (!strlen(raw_redirect_url)) {
+      EmitRedirectWarning(
+          url_, "", "Location header has an empty value");
+      return false;
+    } else {
+      redirect_url = GoogleUrl::Sanitize(raw_redirect_url);
+      raw_redirect_url = redirect_url.c_str();
+    }
+
+    if (redirect_url.find('#') != StringPiece::npos) {
+      EmitRedirectWarning(
+          url_, redirect_url, "Location url has a fragment, not following");
+      return false;
+    }
+
+    scoped_ptr<GoogleUrl> redirect_gurl;
+    redirect_gurl.reset(new GoogleUrl(gurl_, redirect_url));
```

Comment:
No need to heap-allocate this, just GoogleUrl redirect_gurl(gurl_, redirect_url) will do.